script/downsize_uploads.rb

```diff
@@ -77,23 +33,103 @@
     new_file = false
   end
 
+  before = upload.filesize
   upload.filesize = File.size(path)
+  return if upload.filesize > before
+
   upload.sha1 = sha1
   upload.width = w
   upload.height = h
   upload.thumbnail_width = ww
   upload.thumbnail_height = hh
-  next unless upload.save!
 
   if new_file
-    next unless url = Discourse.store.store_upload(File.new(path), upload)
-    next unless upload.update!(url: url)
+    return unless url = Discourse.store.store_upload(File.new(path), upload)
+    upload.url = url
+  end
+
+  if ENV["VERBOSE"]
+    puts "#{original_upload.base62_sha1} -> #{Upload.base62_sha1(sha1)}"
+    puts "#{original_upload.sha1} -> #{sha1}"
+    puts "new file? #{new_file}"
+  end
+
+  upload.save!
+
+  if new_file
+    upload.optimized_images.each(&:destroy!)
+    Discourse.store.remove_upload(original_upload)
+  else
+    begin
+      PostUpload.where(upload_id: original_upload.id).update_all(upload_id: upload.id)
+    rescue ActiveRecord::RecordNotUnique, PG::UniqueViolation
+    end
+
+    User.where(uploaded_avatar_id: original_upload.id).update_all(uploaded_avatar_id: upload.id)
+    UserAvatar.where(gravatar_upload_id: original_upload.id).update_all(gravatar_upload_id: upload.id)
+    UserAvatar.where(custom_upload_id: original_upload.id).update_all(custom_upload_id: upload.id)
+    UserProfile.where(profile_background_upload_id: original_upload.id).update_all(profile_background_upload_id: upload.id)
+    UserProfile.where(card_background_upload_id: original_upload.id).update_all(card_background_upload_id: upload.id)
+    Category.where(uploaded_logo_id: original_upload.id).update_all(uploaded_logo_id: upload.id)
+    Category.where(uploaded_background_id: original_upload.id).update_all(uploaded_background_id: upload.id)
+    CustomEmoji.where(upload_id: original_upload.id).update_all(upload_id: upload.id)
+    ThemeField.where(upload_id: original_upload.id).update_all(upload_id: upload.id)
```

Comment:
can you confirm we have indexes on all these tables (especially user avatar / profile)

Reply:
@SamSaffron `User` and `UserAvatar` have the correct indexes. `UserProfile` needed an update (https://github.com/discourse/discourse/pull/8691). `Category`, `CustomEmoji`, and `ThemeField` don't have indexes on those columns - should I add them?

Reply:
I think we can live with just user and user  avatar and user profile for now, people are only going to rarely run this script so it is not a huge deal if it takes a bit longer